### ISSUE TITLE
"Not allowed to use specific floating IP" makes Cluster enter failure state and stop reconciling (despite later fixing OpenStackCluster)

### ISSUE THREAD

**Issue by MPV** [kubernetes-sigs/cluster-api-provider-openstack#1131]
/kind bug

**What steps did you take and what happened:**

As mentioned in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/1129#issuecomment-1030322290, we upgraded v1alpha3 to v1alpha4 where an `OpenStackCluster` tried to reuse an existing floating IP for the bastion (which it wasn't allowed, but later fixed), causing OpenStackCluster failure to block reconcile of the CAPI Cluster, even after fixing the transient issue of the OpenStackCluster.

Here's an example of the error that we received:
`(rule:create_floatingip and rule:create_floatingip:floating_ip_address) is disallowed by policy`

...leading into:

```json
$ kubectl -n staging-1 get cluster.cluster.x-k8s.io/staging-1-cluster -o json | jq .status
{
  "conditions": [
    {
      "lastTransitionTime": "2022-02-02T15:57:57Z",
      "status": "True",
      "type": "Ready"
    },
    {
      "lastTransitionTime": "2022-02-01T13:36:17Z",
      "status": "True",
      "type": "ControlPlaneInitialized"
    },
    {
      "lastTransitionTime": "2022-02-02T15:57:57Z",
      "status": "True",
      "type": "ControlPlaneReady"
    },
    {
      "lastTransitionTime": "2022-02-02T10:13:02Z",
      "status": "True",
      "type": "InfrastructureReady"
    }
  ],
  "controlPlaneReady": true,
  "failureDomains": {
    "hyperv": {},
    "kvm-linux": {
      "controlPlane": true
    },
    "nova": {}
  },
  "failureMessage": "Failure detected from referenced resource infrastructure.cluster.x-k8s.io/v1alpha4, Kind=OpenStackCluster with name \"staging-1-os-cluster\": failed to associate floating IP with bastion: Expected HTTP response code [200] when accessing [PUT https://myprovider.cloud:9696/v2.0/floatingips/5455a0f7-d1df-4bc9-a576-98fa542cc889], but got 409 instead\n{\"NeutronError\": {\"type\": \"FloatingIPPortAlreadyAssociated\", \"message\": \"Cannot associate floating IP 10.4.93.146 (5455a0f7-d1df-4bc9-a576-98fa542cc889) with port 0095cd34-af54-4786-932b-c583a080e43a using fixed IP 10.223.0.73, as that fixed IP already has a floating IP on external network 620428c3-f24e-4409-81de-7801f4b12802.\", \"detail\": \"\"}}",
  "failureReason": "UpdateError",
  "infrastructureReady": true,
  "observedGeneration": 21,
  "phase": "Failed"
}
```

When CAPO receives this error, even if we were to solve this (by being granted access, or disabling the bastion altogether using #1076), since CAPO causes the `Cluster` to get `status.failureMessage`, it won't keep reconciling anymore (even if the `OpenStackCluster` wouldn't have this issue anymore).

Here's an example of how the `OpenStackCluster` isn't faulty anymore (but with the `Cluster` above still staying in Failed phase):

```json
$ kubectl -n staging-1 get openstackclusters.infrastructure.cluster.x-k8s.io staging-1-os-cluster -o json | jq .status
{
  "externalNetwork": {
    "id": "620428c3-f24e-4409-81de-7801f4b12802",
    "name": "cuspageroonline_lb-net_530"
  },
  "failureDomains": {
    "hyperv": {},
    "kvm-linux": {
      "controlPlane": true
    },
    "nova": {}
  },
  "network": {
    "apiServerLoadBalancer": {
      "id": "83aefff5-3416-4f5b-94da-f12817dec070",
      "internalIP": "10.223.3.13",
      "ip": "10.4.93.174",
      "name": "k8s-clusterapi-cluster-staging-1-staging-1-cluster-kubeapi"
    },
    "id": "23f96472-d1c3-4108-aba7-a73fae0d44f3",
    "name": "k8s-clusterapi-cluster-staging-1-staging-1-cluster",
    "router": {
      "id": "418a3ab3-ce35-496a-926e-e27c267368b9",
      "name": "k8s-clusterapi-cluster-staging-1-staging-1-cluster"
    },
    "subnet": {
      "cidr": "10.223.0.0/18",
      "id": "83ab93f1-8ea6-4c46-925b-50ff19fef92f",
      "name": "k8s-clusterapi-cluster-staging-1-staging-1-cluster"
    }
  },
  "ready": true
}
```

**What did you expect to happen:**

CAPI Cluster being able to reconcile again after CAPO OpenStackCluster reconciles successfully again.

**Anything else you would like to add:**
Very much related to #1116 (but in that case it's `OpenStackMachine` affected by transient failures, but here it's a transient failure in `OpenStackCluster` causing `Cluster` to enter an error state and stop reconciling.

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): `d76e19e8e0688552eb662ed499cd6a25713716e8`
- Cluster-API version: `v1.0.2`
- OpenStack version: N/A
- Kubernetes version (use `kubectl version`): N/A

**Comment by MPV** [comment]
Also curious how other CAPx providers have solved this. Presumably CAPO isn't the first to reach this issue.

**Comment by jichenjc** [comment]
>>>Firstly, I don't think we should ever set the Cluster to failed due to a failure of the bastion specifically, because the bastion is ephemeral. That's a bug and we should fix it.

+1 to this 

>>>Is there any way to create a tighter and more obvious coupling between validation logic in the webhook and special cases like this where the validation allows the user to fix certain failures

I think the webhook only focus on whether `update` is allowed on some fields and the logic things such as `whether some param input as resource uuid can be found ` is not something should be done in webhook

**Comment by jichenjc** [comment]
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1138 created

I think we might have more other cases fit for this 
but this need another thought/discussion to make change...

**Comment by MPV** [comment]
Would #1138 remove a previous such failure too?

Asking since we have a cluster that is stuck in failed mode and would prefer it to heal than need recreation after upgrading.

**Comment by seanschneeweiss** [comment]
> Would #1138 remove a previous such failure too?
> 
> Asking since we have a cluster that is stuck in failed mode and would prefer it to heal than need recreation after upgrading.

For now you probably have to patch the status of the cluster yourself if stuck with a failureReason or failureMessage.

**Comment by mdbooth** [comment]
Sounds like this is fixed by #1138. https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1153 will also improve this.

**Comment by MPV** [comment]
Yup, and we ended up fixing it for now using the workaround of editing the raw status of the cluster. 👍
